packages/babel-plugin-transform-block-scoping/src/loop.ts

```diff
@@ -0,0 +1,334 @@
+import { template, types as t } from "@babel/core";
+import type { NodePath, Visitor, Binding } from "@babel/traverse";
+
+interface LoopBodyBindingsState {
+  blockScoped: Binding[];
+}
+
+const collectLoopBodyBindingsVisitor: Visitor<LoopBodyBindingsState> = {
+  "Expression|Declaration|Loop"(path) {
+    path.skip();
+  },
+  Scope(path, state) {
+    if (path.isFunctionParent()) path.skip();
+
+    const { bindings } = path.scope;
+    for (const name of Object.keys(bindings)) {
+      const binding = bindings[name];
+      if (
+        binding.kind === "let" ||
+        binding.kind === "const" ||
+        binding.kind === "hoisted"
+      ) {
+        state.blockScoped.push(binding);
+      }
+    }
+  },
+};
+
+export function getLoopBodyBindings(loopPath: NodePath<t.Loop>) {
+  const state: LoopBodyBindingsState = { blockScoped: [] };
+  loopPath.traverse(collectLoopBodyBindingsVisitor, state);
+  return state.blockScoped;
+}
+
+export function getUsageInBody(binding: Binding, loopPath: NodePath<t.Loop>) {
+  // UpdateExpressions are counted both as a reference and a mutation,
+  // so we need to de-duplicate them.
+  const seen = new WeakSet<t.Node>();
+
+  let capturedInClosure = false;
+
+  const constantViolations = filterMap(binding.constantViolations, path => {
+    const { inBody, inClosure } = relativeLoopLocation(path, loopPath);
+    if (!inBody) return null;
+    capturedInClosure ||= inClosure;
+
+    const id = path.isUpdateExpression()
+      ? path.get("argument")
+      : path.isAssignmentExpression()
+      ? path.get("left")
+      : null;
+    if (id) seen.add(id.node);
+    return id as NodePath<t.Identifier> | null;
+  });
+
+  const references = filterMap(binding.referencePaths, path => {
+    if (seen.has(path.node)) return null;
+
+    const { inBody, inClosure } = relativeLoopLocation(path, loopPath);
+    if (!inBody) return null;
+    capturedInClosure ||= inClosure;
+
+    return path as NodePath<t.Identifier>;
+  });
+
+  return {
+    capturedInClosure,
+    hasConstantViolations: constantViolations.length > 0,
+    usages: references.concat(constantViolations),
+  };
+}
+
+function relativeLoopLocation(path: NodePath, loopPath: NodePath<t.Loop>) {
+  const bodyPath = loopPath.get("body");
+  let inClosure = false;
+
+  for (let currPath = path; currPath; currPath = currPath.parentPath) {
+    if (currPath.isFunction()) inClosure = true;
```

Comment:
Q: Should we exclude computed keys for methods? Should we also treat class initializers and static blocks as `inClosure`?

Reply:
I included classes.

I could exclude computed keys so that this code doesn't need a closure:
```js
for (let i;;) {
  class A { [i]() {} }
}
```
because `i` is eagerly accessed, but it complicates the code for little benefit (it's very rare to use the loop variable in a computed key).